### PR TITLE
Modify Threading Scheme to Improve Performance with Large Numbers of Threads

### PR DESCRIPTION
**Pull request by Mr4k**
### Checklist

* [x] I have read the [Contributor Guide](../../CONTRIBUTING.md)
* [x] I have read and agree to the [Code of Conduct](../../CODE_OF_CONDUCT.md)
* [x] I have added a description of my changes and why I'd like them included in the section below

### Description of Changes

Disclaimer: This pr is a big proposed change and is not fully ready yet (inpaint and tiling are currently disabled but should be easily fixable). The code isn't the cleanest (feel free to let me know if it's too hard to read) and could be optimized further but I wanted to put it up as a proof of concept see what you think, make sure my big assumptions about how the code works were not very wrong, and see how this runs on other people's machines before I go further.

Motivation:
After seeing that @h3r2tic had virtually no improvements from my previous pr on their 32 core cpu I rented a larger cloud machine (64 logical cores, Intel Xeon) to test out how my previous changes behaved with many cores. It turns out I didn't see any improvement either. I then tried with lower numbers of threads (48, 32, 24, 16, 8, etc) and noticed that the run time actually seemed to improve as I decreased the number of threads (after some fiddling I found 22 threads was the best number for my instance) and barely went up until < 16 threads were in use. I figured this was a contention problem, so I added some additional logging to measure the time it took to acquire various locks and found (at least on that machine) the time it took to acquire a write lock on the rtree went up proportionally with the number of threads. Also as a side effect I believe it dwarfed any improvement from my previous change to the point where it was barely noticeable.

Proposed Improvements:
I've made several changes which I have seen reduce contention on my test machines.

1. Replace the RTree with a grid of Rtrees
Advantages:
Most of the time (at least in the later stages of synthesis) two threads may be working on totally different areas of the image and will not need to access the same sets of pixels and therefore should not have to wait to read / write. To address this after the first few stages of synthesis I replace the single rtree with a grid of rtrees. Writes only lock the rtree in the grid cell which the pixel is in and reads just read from a cell and its neighbors. Overall this has seemed to improve contention considerably in my tests. (Possibly a multi threaded rtree is better than this but I couldn't find one and it seemed more difficult / error prone to build for likely little upside)
Drawbacks:
The actual read itself is a little slower, this is balanced out by the fact that far more threads can read and write at the same time but it should be optimized a little more for machines with fewer cores
Technically we can miss some neighbors if a pixels k nearest neighbors are outside of the current cell + 8 adjacent cells, however because we only use a grid after the first few steps there should be enough pixels where this is never a problem. I have not seen quality degrade (see below)
Places for improvement:
The read can definitely be optimized more (especially important for cpus with small numbers of cores)
The number of grid cells is currently just a random constant but really should vary with the number of threads / size of the image / k

2. Don’t write to resolved queue until after stage is complete (don't updated it when a pixel is placed)
After I switched to the rtree grid I noticed that although writes were really fast the resolved queue seemed to be suddenly very hard to acquire a lock on. After looking at the code more I decided that each thread could have its own resolved queue and I could combine them at the end of the stage. I don't think this actually changes the behavior of the code because the only time we read from the resolved queue is to redo pixels done in the previous stages so we don't care what pixels other threads have updated in the current stage.
Advantages:
The resolved queue is no longer backed up leading to a speed up
Disadvantages:
None that I see

3. Remove the Update Queue entirely
I removed the update queue to simplify things a little because the writes were now fast enough where I could write every time instead of needing to batch.
Advantages:
Code is simpler, very small speed up
Disadvantages:
I can't think of too much but possibly this could be bad not to have in the first few stages before we switch to the grid of rtrees
Could also be worse for smaller grid sizes

Results:
I’ve tested on only two large machines so far:
Google Cloud N1:
Threads: 64
Cpu: Intel(R) Xeon(R) CPU @ 2.30GHz:
```
Command: time cargo run --release -- --threads 64 --out-size 2048 --out out/01.jpg generate imgs/1.jpg
BEFORE: ~1m16s
AFTER: ~28.8s

Command: time cargo run --release -- --threads 32 --out-size 2048 --out out/01.jpg generate imgs/1.jpg
BEFORE: ~1m12.8s
AFTER: ~35.7s

Command: time cargo run --release -- --threads 64 --out-size 1024 --out out/01.jpg generate imgs/1.jpg
BEFORE: ~17.5s
AFTER: ~7.5s

Command time cargo run --release -- --threads 64 --out-size 512 --out out/01.jpg generate imgs/1.jpg
BEFORE: ~4.1s
AFTER: ~1.9s
```

AWS Z1d:
Threads: 48
Cpu: Intel® Xeon® Scalable (Clock speed is a little unclear but Amazon claims up to 4ghz) 
```
Command: time cargo run --release -- --threads 48  --out-size 2048 --out out/01.jpg generate imgs/1.jpg
BEFORE: ~41.5 seconds
AFTER: ~22.5 seconds

Command: time cargo run --release --  --threads 24  --out-size 2048 --out out/01.jpg generate imgs/1.jpg
BEFORE: ~44.5 seconds
AFTER: ~29 seconds

Command: time cargo run --release --  --threads 48 --out-size 1024 --out out/01.jpg generate imgs/1.jpg
BEFORE: ~9.7 seconds
AFTER: ~5.7 seconds 

time cargo run --release --  --threads 48 --out-size 512 --out out/01.jpg generate imgs/1.jpg
BEFORE: ~2.36 seconds
AFTER: ~1.47 seconds
```

I’m also curious to see what others get on cpus with large numbers of cores. @h3r2tic especially want to know how this runs on your 32 core cpu if you have the time.

Image Quality:
Because this change affects the threads and is fairly major I wanted to make sure that image quality did not change. It looks like tests still pass for everything that I didn’t break so the image hashes must be fairly close. I also manually inspected some large images but did not see any major changes. However it is important that the threads share one rtree for the first few stages before there are enough pixels to fill out a decent portion of the grid cells.

### Related Issues

I don’t believe this is related to any mentioned issue

### CODE REVIEW

**Comment by Mr4k** [comment]
Sped up the read from the grid of rtrees which brought the runtime down a little:
From a test Google Cloud Machine with 64 threads
```
Command: time cargo run --release -- --out-size 1024  --out out/01.jpg generate imgs/1.jpg
Before grid read speed up:
~7.5 seconds
After grid read speed up:
~5.6 seconds

Command: time cargo run --release -- --out-size 2048 --out out/01.jpg generate imgs/1.jpg
Before grid read speed up:
~28 seconds
After grid read speed up:
~24 seconds
```
But most importantly it now runs a lot faster cpus with few threads.
My development macbook (4 logical cores):
```
Command: time cargo run --release -- --out out/01.jpg generate imgs/1.jpg
Before: ~20s
After: ~11.5s
```

File: lib/src/multires_stochastic_texture_synthesis.rs [hunk 1]
```diff
@@ -1263,16 +1263,17 @@ impl STree {
...
+            (chunk_x, chunk_y, true, 0, 0),
+            (chunk_x + 1, chunk_y, false, (chunk_x + 1) * self.chunk_size as i32, y as i32),
+            (chunk_x - 1, chunk_y, false, chunk_x * self.chunk_size as i32, y as i32),
+            (chunk_x, chunk_y - 1, false, x as i32, chunk_y * self.chunk_size as i32),
+            (chunk_x, chunk_y + 1, false, x as i32, (chunk_y + 1) * self.chunk_size as i32),
+            (chunk_x + 1, chunk_y + 1, false, (chunk_x + 1) * self.chunk_size as i32, (chunk_y + 1) * self.chunk_size as i32),
+            (chunk_x - 1, chunk_y + 1, false, chunk_x * self.chunk_size as i32, (chunk_y + 1) * self.chunk_size as i32),
+            (chunk_x + 1, chunk_y - 1, false, (chunk_x + 1) * self.chunk_size as i32, chunk_y * self.chunk_size as i32),
+            (chunk_x - 1, chunk_y - 1, false, chunk_x * self.chunk_size as i32, chunk_y * self.chunk_size as i32),
         ];
```
Comment (Jake-Shadle) [inline lib/src/multires_stochastic_texture_synthesis.rs]: I would really prefer this to be a `struct` defined inside the the function, tuples are fine for prototyping but end up being really tedious to use over the long term, especially > 2 length.

**Comment by h3r2tic** [comment]
That's an impressive speed up! I'll have a look :)

File: lib/src/multires_stochastic_texture_synthesis.rs (hunk 1 above)
Comment (Mr4k) [reply lib/src/multires_stochastic_texture_synthesis.rs]: Thanks for the pointer! This code is still pretty prototype-y. I will make a cleanup pass later and can definitely change this.

**Comment by Jake-Shadle** [comment]
Oh and FYI, if you are doing your iterative benchmarking via the cli you might want to check out hyperfine.

**Comment by Mr4k** [comment]
Had a little time so I did some code clean up. Inpaint is now reenabled, tiling is still broken but should be a simple fix. Also @h3r2tic wanted to see if you ever got a chance to benchmark these changes on your hardware yet (no worries if not, I don't want to get in the way of your actual work)

File: lib/src/multires_stochastic_texture_synthesis.rs (hunk 1 above)
Comment (Mr4k) [reply lib/src/multires_stochastic_texture_synthesis.rs]: I changed these tuples to a in function struct in my most recent commit

**Comment by Mr4k** [comment]
Fixed in painting. This pr is not mergable yet because it does not work for rectangular images but I will add that likely tomorrow

File: lib/src/multires_stochastic_texture_synthesis.rs [hunk 2]
```diff
@@ -762,6 +704,30 @@ impl Generator {
...
+            // if we already have resolved pixels from an inpaint or multiexample add them to this tree grid
+            let resolved_queue = &mut self.resolved.write().unwrap();
+            let pixels_to_update: Vec<([i32; 2], CoordFlat, Score)> = resolved_queue
+                .drain(..)
+                .map(|a| {
+                    let coord_2d = a.0.to_2d(self.output_size);
+                    ([coord_2d.x as i32, coord_2d.y as i32], a.0, a.1)
+                })
+                .collect();
+            self.flush_resolved(
```
Comment (Mr4k) [inline lib/src/multires_stochastic_texture_synthesis.rs]: Note this changes behavior for inpainting + tiling. Now pixels which are not part of the inpaint region will be taken into account when tiling. I could change this back to the old behavior (where those pixels were ignored) by imho this behavior seem correct

**Comment by Mr4k** [comment]
Functionality wise this pr is now complete.
Also took a quick performance benchmark using hyperfine:
Google Cloud 64 core cpu 
```
Before: 
Benchmark #1: ./target/release/texture-synthesis --out-size 2048 --out out/01.jpg generate imgs/1.jpg
  Time (mean ± σ):     64.681 s ±  0.256 s    [User: 734.705 s, System: 104.357 s]
  Range (min … max):   64.293 s … 65.154 s    10 runs

After:
Benchmark #1: ./target/release/texture-synthesis --out-size 2048 --out out/01.jpg generate imgs/1.jpg
  Time (mean ± σ):     20.544 s ±  0.182 s    [User: 796.544 s, System: 9.745 s]
  Range (min … max):   20.362 s … 20.999 s    10 runs
```

File: lib/src/multires_stochastic_texture_synthesis.rs [hunk 3]
```diff
@@ -761,6 +702,35 @@ impl Generator {
...
+        {
+            // now that we have all of the parameters we can setup our initial tree grid
+            let tile_adjusted_width = (self.output_size.width as f32 * 1.1) as u32 + 1;
+            let tile_adjusted_height = (self.output_size.height as f32 * 1.1) as u32 + 1;
+            self.tree_grid = TreeGrid::new(
+                tile_adjusted_width,
+                tile_adjusted_height,
+                max(tile_adjusted_width, tile_adjusted_height),
+                (self.output_size.width as f32 * 0.05) as u32 + 1,
+                (self.output_size.height as f32 * 0.05) as u32 + 1,
```
Comment (h3r2tic) [inline lib/src/multires_stochastic_texture_synthesis.rs]: What's the reasoning behind the `1.1` multiplier, and why is `1` added on top of it?

File: lib/src/multires_stochastic_texture_synthesis.rs (hunk 2 above)
Comment (h3r2tic) [reply lib/src/multires_stochastic_texture_synthesis.rs]: Perhaps @anopara could have a look. I'm not quite sure what the implications are with the previous and the new behaviors.

File: lib/src/multires_stochastic_texture_synthesis.rs [hunk 4]
```diff
@@ -761,6 +702,35 @@ impl Generator {
         let l2_precomputed = PrerenderedU8Function::new(metric_l2);
         let mut total_processed_pixels = 0;
         let max_workers = params.max_thread_count;
+        let mut has_fanned_out = false;
```
Comment (h3r2tic) [inline lib/src/multires_stochastic_texture_synthesis.rs]: Can you add a comment to clarify the "fanned out"? Perhaps something along the lines of "use a single R*-tree initially, and fan out to a grid thereof later".

File: lib/src/multires_stochastic_texture_synthesis.rs [hunk 5]
```diff
@@ -817,6 +813,10 @@ impl Generator {
                         let max_candidate_count = params.nearest_neighbors as usize
                             + params.random_sample_locations as usize;
 
+                        let my_thread_num = thread_counter.fetch_add(1, Ordering::Relaxed);
```
Comment (h3r2tic) [inline lib/src/multires_stochastic_texture_synthesis.rs]: Nitpick: should be `my_thread_id` for consistency with other ordinals.

File: lib/src/multires_stochastic_texture_synthesis.rs [hunk 6]
```diff
@@ -981,8 +986,13 @@ impl Generator {
             })
             .unwrap();
 
-            // Some items might still be pending a resolve flush. Do it now before we start the next stage.
-            self.force_flush_resolved(is_tiling_mode);
+            {
+                // append all resolved threads to resolved
```
Comment (h3r2tic) [inline lib/src/multires_stochastic_texture_synthesis.rs]: Sounds like we're appending threads. Would rephrase to "append all per-thread resolved lists to the global list"

File: lib/src/multires_stochastic_texture_synthesis.rs [hunk 7]
```diff
@@ -981,8 +986,13 @@ impl Generator {
...
+                let mut resolved = self.resolved.write().unwrap();
+                for thread_resolved in pixels_resolved_this_stage {
+                    resolved.extend(thread_resolved.lock().unwrap().iter());
```
Comment (h3r2tic) [inline lib/src/multires_stochastic_texture_synthesis.rs]: Might be a bit clearer like this:

```suggestion
                    resolved.append(&mut thread_resolved.into_inner().unwrap());
```

File: lib/src/multires_stochastic_texture_synthesis.rs [hunk 8]
```diff
@@ -1191,6 +1201,213 @@ impl PrerenderedU8Function {
...
+    rtrees: Vec<RwLock<RTree<[i32; 2]>>>,
+}
+
+// This is a grid of rtrees
+// The idea is that most pixels after the first couple steps will have their neighbors close by
+impl TreeGrid {
+    pub fn new(width: u32, height: u32, chunk_size: u32, offset_x: u32, offset_y: u32) -> TreeGrid {
+        let mut rtrees: Vec<RwLock<RTree<[i32; 2]>>> = Vec::new();
+        let grid_width = width / chunk_size + 1;
+        let grid_height = height / chunk_size + 1;
```
Comment (h3r2tic) [inline lib/src/multires_stochastic_texture_synthesis.rs]: If you want round-up behavior here, it should probably be written as `(a + (b - 1)) / b`, e.g. `(width + chunk_size - 1) / chunk_size`. With `a / b + 1`, you get an extra element when `a` is divisible by `b`.

File: lib/src/multires_stochastic_texture_synthesis.rs [hunk 9]
```diff
@@ -1191,6 +1201,213 @@ impl PrerenderedU8Function {
...
+            grid_width,
+            grid_height,
+            offset_x: offset_x as i32,
+            offset_y: offset_y as i32,
+            chunk_size,
+        }
+    }
+
+    #[inline]
+    fn get_space(&self, x: u32, y: u32) -> usize {
```
Comment (h3r2tic) [inline lib/src/multires_stochastic_texture_synthesis.rs]: Why call this `get_space`? Considering how you invoke it in the code, e.g.:

`let my_tree_index = self.get_space`,

perhaps it should be called `get_tree_index`.

File: lib/src/multires_stochastic_texture_synthesis.rs [hunk 10]
```diff
@@ -1191,6 +1201,213 @@ impl PrerenderedU8Function {
...
+// This is a grid of rtrees
+// The idea is that most pixels after the first couple steps will have their neighbors close by
+impl TreeGrid {
+    pub fn new(width: u32, height: u32, chunk_size: u32, offset_x: u32, offset_y: u32) -> TreeGrid {
+        let mut rtrees: Vec<RwLock<RTree<[i32; 2]>>> = Vec::new();
+        let grid_width = width / chunk_size + 1;
+        let grid_height = height / chunk_size + 1;
+        rtrees.resize_with((grid_width * grid_height) as usize, || {
+            RwLock::new(RTree::new())
+        });
```
Comment (h3r2tic) [inline lib/src/multires_stochastic_texture_synthesis.rs]: Could just use `let rtrees = vec![RwLock::new(RTree::new()); grid_width * grid_height];`

File: lib/src/multires_stochastic_texture_synthesis.rs [hunk 11]
```diff
@@ -1191,6 +1201,213 @@ impl PrerenderedU8Function {
...
+            chunk_size,
+        }
+    }
+
+    #[inline]
+    fn get_space(&self, x: u32, y: u32) -> usize {
+        (x * self.grid_height + y) as usize
+    }
+
+    pub fn force_insert(&self, x: i32, y: i32) {
```
Comment (h3r2tic) [inline lib/src/multires_stochastic_texture_synthesis.rs]: Why "force" insert? I don't see a non-forceful version, and it's not clear what's forceful about this one.

File: lib/src/multires_stochastic_texture_synthesis.rs [hunk 12]
```diff
@@ -1191,6 +1201,213 @@ impl PrerenderedU8Function {
...
+        struct ChunkSearchInfo {
+            x: i32,
+            y: i32,
+            center: bool,
+            closest_point_on_boundary_x: i32,
+            closest_point_on_boundary_y: i32,
+        }
+
+        // Assume that all k nearest neighbors are in these cells
+        // it looks like we are rarely wrong once enough pixels are filled in
```
Comment (h3r2tic) [inline lib/src/multires_stochastic_texture_synthesis.rs]: I don't think I follow the comment. Is this to indicate that locking all cells at once results in deadlocks?

File: lib/src/multires_stochastic_texture_synthesis.rs [hunk 13]
```diff
@@ -1191,6 +1201,213 @@ impl PrerenderedU8Function {
...
+            },
+        ];
+        // Note locking all of them at different times is fine as opposed to
+        // all at once but this behavior should be noted
+        let mut tmp_result: Vec<(i32, i32, i32)> = Vec::with_capacity(k * 9);
+        result.clear();
+        result.reserve(k);
+
+        // this isn't really the kth best distance but should be good enough for how simple it is
+        let mut kth_best_squared_distance = i32::max_value();
```
Comment (h3r2tic) [inline lib/src/multires_stochastic_texture_synthesis.rs]: Using `i32` here starts feeling a bit tight -- some people use _giant_ images, and there's a chance of this overflowing, particularly in the first stage of processing before "fanning out" to a larger grid of trees.

File: lib/src/multires_stochastic_texture_synthesis.rs [hunk 14]
```diff
@@ -1191,6 +1201,213 @@ impl PrerenderedU8Function {
...
+                && place_to_look.x < self.grid_width as i32
+                && place_to_look.y >= 0
+                && place_to_look.y < self.grid_height as i32
+            {
+                let is_center = place_to_look.center;
+
+                // a tiny optimization to help us throw far away neighbors
+                // saves us a decent amount of reads
+                if !is_center {
+                    let squared_distance_to_closest_possible_point_on_chunk = (x as i32
```
Comment (h3r2tic) [inline lib/src/multires_stochastic_texture_synthesis.rs]: Ditto, a bunch of potentially dangerous `i32` here.

File: lib/src/multires_stochastic_texture_synthesis.rs [hunk 15]
```diff
@@ -1191,6 +1201,213 @@ impl PrerenderedU8Function {
...
+        for place_to_look in places_to_look.iter() {
+            if place_to_look.x >= 0
+                && place_to_look.x < self.grid_width as i32
+                && place_to_look.y >= 0
+                && place_to_look.y < self.grid_height as i32
+            {
+                let is_center = place_to_look.center;
+
+                // a tiny optimization to help us throw far away neighbors
+                // saves us a decent amount of reads
```
Comment (h3r2tic) [inline lib/src/multires_stochastic_texture_synthesis.rs]: Is this actually correct though? This seems to be just point-to-point distance against `closest_point_on_boundary_`. The latter seem to be centered on edges of cells, but are not guaranteed to be the closest point inside the cell.

File: lib/src/multires_stochastic_texture_synthesis.rs [hunk 16]
```diff
@@ -1191,6 +1201,213 @@ impl PrerenderedU8Function {
...
+            x: i32,
+            y: i32,
+            center: bool,
+            closest_point_on_boundary_x: i32,
+            closest_point_on_boundary_y: i32,
+        }
+
+        // Assume that all k nearest neighbors are in these cells
+        // it looks like we are rarely wrong once enough pixels are filled in
+        let places_to_look = vec![
```
Comment (h3r2tic) [inline lib/src/multires_stochastic_texture_synthesis.rs]: Looks like we can avoid a heap allocation by just removing the `vec!` characters :)

File: lib/src/multires_stochastic_texture_synthesis.rs [hunk 17]
```diff
@@ -1191,6 +1201,213 @@ impl PrerenderedU8Function {
...
+                            (
+                                (*a)[0],
+                                (*a)[1],
+                                ((*a)[0] - x as i32) * ((*a)[0] - x as i32)
+                                    + ((*a)[1] - y as i32) * ((*a)[1] - y as i32),
+                            )
+                        }),
+                );
+
+                // this isn't really the kth best distance but it's an okay approximation
```
Comment (h3r2tic) [inline lib/src/multires_stochastic_texture_synthesis.rs]: I think a good deal of "this isn't really X" could be avoided by just calling the variable differently. How about something like `kth_nearest_distance_upper_bound`?

File: lib/src/multires_stochastic_texture_synthesis.rs (hunk 3 above)
Comment (Mr4k) [reply lib/src/multires_stochastic_texture_synthesis.rs]: This was just a low quality thing on my part. The 1.1 is because in tiling mode you have boundaries which are 5% of the image size. I will move that into a constant. You are also right about the +1 being for rounding up. I will follow your suggestion below.

File: lib/src/multires_stochastic_texture_synthesis.rs (hunk 4 above)
Comment (Mr4k) [reply lib/src/multires_stochastic_texture_synthesis.rs]: Will do!

File: lib/src/multires_stochastic_texture_synthesis.rs (hunk 5 above)
Comment (Mr4k) [reply lib/src/multires_stochastic_texture_synthesis.rs]: 👍

File: lib/src/multires_stochastic_texture_synthesis.rs (hunk 6 above)
Comment (Mr4k) [reply lib/src/multires_stochastic_texture_synthesis.rs]: haha true I'll rewrite that

File: lib/src/multires_stochastic_texture_synthesis.rs (hunk 8 above)
Comment (Mr4k) [reply lib/src/multires_stochastic_texture_synthesis.rs]: great suggestion will do!

File: lib/src/multires_stochastic_texture_synthesis.rs (hunk 9 above)
Comment (Mr4k) [reply lib/src/multires_stochastic_texture_synthesis.rs]: 👍

File: lib/src/multires_stochastic_texture_synthesis.rs (hunk 10 above)
Comment (Mr4k) [reply lib/src/multires_stochastic_texture_synthesis.rs]: 👍

File: lib/src/multires_stochastic_texture_synthesis.rs (hunk 11 above)
Comment (Mr4k) [reply lib/src/multires_stochastic_texture_synthesis.rs]: This is an artifact from before I removed the update queue when not every write inserted immediately. I will now just name it insert.

File: lib/src/multires_stochastic_texture_synthesis.rs (hunk 12 above)
Comment (Mr4k) [reply lib/src/multires_stochastic_texture_synthesis.rs]: I don't think this has to do with deadlocks.
Basically in my mind there were two ways to do this function:
1. Lock all the cells we are reading from at once so no new pixels can be added to any of them before the read is done
2. Lock each cell one at a time which may mean a new results pop up in between cell reads I went with this approach. I don't think i's bad but I wanted to make a note about the choice because it seemed subtle.
I'll rewrite this

File: lib/src/multires_stochastic_texture_synthesis.rs (hunk 13 above)
Comment (Mr4k) [reply lib/src/multires_stochastic_texture_synthesis.rs]: oh wow never considered an image of that size. I'll bump to i64

File: lib/src/multires_stochastic_texture_synthesis.rs (hunk 14 above)
Comment (Mr4k) [reply lib/src/multires_stochastic_texture_synthesis.rs]: 👍

File: lib/src/multires_stochastic_texture_synthesis.rs (hunk 15 above)
Comment (Mr4k) [reply lib/src/multires_stochastic_texture_synthesis.rs]: Sorry the name there might be a little confusing what I mean by closet point in cell is closest point on boundary. The optimization works as follows:
diagram
If the green dot is the pixel we are talking about and we know that the k nearest neighbors are all within the purple circle then we don't have to search any of the red cells because the closest possible points (the boundary points) to our pixel are further than any of our k closest neighbors.

File: lib/src/multires_stochastic_texture_synthesis.rs (hunk 12 above)
Comment (h3r2tic) [reply lib/src/multires_stochastic_texture_synthesis.rs]: I think 2. sounds okay, and will be less susceptible to deadlocks. I think with 1. deadlocks would be very likely, as multiple threads could lock subsets of cells, and get stuck trying to acquire all of them.

File: lib/src/multires_stochastic_texture_synthesis.rs (hunk 16 above)
Comment (Mr4k) [reply lib/src/multires_stochastic_texture_synthesis.rs]: Thanks good catch!

File: lib/src/multires_stochastic_texture_synthesis.rs (hunk 17 above)
Comment (Mr4k) [reply lib/src/multires_stochastic_texture_synthesis.rs]: 👍

**Comment by Mr4k** [comment]
Thanks so much for reviewing! I'll make the changes soon and generally try to make the naming better! And of course if you get a chance I'd love to know runtimes on more reference rigs but no pressure if you're busy.

File: lib/src/multires_stochastic_texture_synthesis.rs (hunk 2 above)
Comment (anopara) [reply lib/src/multires_stochastic_texture_synthesis.rs]: Hi, @Mr4k. First of all, this is some awesome work, thank you! Could you please elaborate on "Now pixels which are not part of the inpaint region will be taken into account when tiling" 😅 I'm not quite sure how exactly they will be used during tiling?

File: lib/src/multires_stochastic_texture_synthesis.rs (hunk 12 above)
Comment (Mr4k) [reply lib/src/multires_stochastic_texture_synthesis.rs]: Yes sorry I meant deadlock as a situation where no forward progress was possible. The large advantage to 2 is that faster to lock them one at a time as you said

File: lib/src/multires_stochastic_texture_synthesis.rs (hunk 2 above)
Comment (Mr4k) [reply lib/src/multires_stochastic_texture_synthesis.rs]: Hi @anopara. Thanks for making this awesome project! Here is an example of what I'm talking about. Consider the output of the following command.
```
cargo run --release -- --out-size 304x308 --inpaint circles_mask.png --tiling -o imgs/01.png generate circles.png
```
using the following images:
circles.png:
circles
circles_mask.png:
circles_mask

On master this generate this following image:
without_tiling_inpaint

On my branch this will generate this image:
with_tiling_inpaint

What I mean is that in my image the inpaint takes into account the pixels on the other side of the image to make it tile. As far as I understand this is because in master when you add the static pixels which are not part of the inpaint region to the rtree (https://github.com/EmbarkStudios/texture-synthesis/blob/897d10aaf41efde9acf4122c7c6d19629a0425d6/lib/src/multires_stochastic_texture_synthesis.rs#L265) you do not "reflect" them across image borders like you do when you add pixels during synthesis (https://github.com/EmbarkStudios/texture-synthesis/blob/897d10aaf41efde9acf4122c7c6d19629a0425d6/lib/src/multires_stochastic_texture_synthesis.rs#L295). What I've done is just to use the same function for both which provides mirroring as a side effect.

Comment (anopara) [reply lib/src/multires_stochastic_texture_synthesis.rs]: Ah! That is a very good observation and that is indeed correct. Thank you for spotting this case. I guess, we never tested with/thought of tiling to be generated only on one border 😅 Your proposed change is definitely a more desirable behavior! 👍

**Comment by h3r2tic** [comment]
Tested on my Threadripper! Pretty consistent with the "Google Cloud 64 core cpu".

```
BEFORE:

Benchmark #1: target\release\texture-synthesis --out-size 2048 --out out/01.jpg generate imgs/1.jpg
  Time (mean ± σ):     62.349 s ±  0.255 s    [User: 1.4 ms, System: 4.4 ms]
  Range (min … max):   61.825 s … 62.664 s    10 runs

AFTER:

Benchmark #1: target\release\texture-synthesis --out-size 2048 --out out/01.jpg generate imgs/1.jpg
  Time (mean ± σ):     23.737 s ±  0.488 s    [User: 3.0 ms, System: 3.6 ms]
  Range (min … max):   23.034 s … 24.373 s    10 runs
```

**Comment by repi** [comment]
Impressive improvement!

File: lib/src/multires_stochastic_texture_synthesis.rs (hunk 15 above)
Comment (h3r2tic) [reply lib/src/multires_stochastic_texture_synthesis.rs]: Could you not end up in the following situation?

image

I have highlighted in blue what I'm assuming to be the boundary points of the missed cells.

Comment (Mr4k) [reply lib/src/multires_stochastic_texture_synthesis.rs]: Thanks for the picture! I think the position of the blue points is incorrect. They should really be the closest point on each boundary to the current pixel. For the 4 corner cells this point is always a constant (the closest corner on that cell to the center of the grid) no matter which pixel we are addressing in the middle. However for the cells you highlighted as problematic the boundary points I compute are actually based on the green pixels x or y coordinates (in the case of your picture one would be right below the green dot and the other would be directly to the right)

Comment (Mr4k) [reply lib/src/multires_stochastic_texture_synthesis.rs]: Here's a revised version of your picture
diagram_with_projected_points

File: lib/src/multires_stochastic_texture_synthesis.rs (hunk 10 above)
Comment (Mr4k) [reply lib/src/multires_stochastic_texture_synthesis.rs]: Unfortunately got an error trying to do this saying that RwLock<Rtree> does not implement the copy constructor

File: lib/src/multires_stochastic_texture_synthesis.rs (hunk 15 above)
Comment (h3r2tic) [reply lib/src/multires_stochastic_texture_synthesis.rs]: Aha, so I did misread the code, thinking you were offsetting from the center. Make sense now, thanks!

**Comment by Mr4k** [comment]
Good to hear! (Although I did accidentally just add and remove a print statement so I think I'll need another approval)

Also just for completeness's sake here are some things I thought of that I decided were out of scope of this pr but might be worth mentioning:
- I'm not sure how common this use case is but for inpainting a few small sparse regions on a large image an sparse version of the tree grid may provide better performance
- In my benchmarking I have noticed that since the contention has been reduced the single threaded first few stages for multithreaded generation (especially with many threads) has become a large bottleneck (up to 50% of the time in some case iirc). A potentially future improvement may be see if that can be some what parallelized especially for larger images
- Also after benchmarking I was also left wondering if an rstar tree is the ideal data structure for this kind of problem (now takes up about 50% of the generation step for a pixel iirc) given that most of the time your k nearest neighbors will be very close to you in predictable positions. I'm not sure what a better data structure looks like but it was just a thing I thought was interesting

Also finally thanks so much @h3r2tic @anopara @Jake-Shadle for taking the time to benchmark / review my code! It was really fun to work on this.

**Comment by h3r2tic** [comment]
Good thoughts!

Like you say, the initial serial part of the algorithm will certainly suffer from Amdahl's law. I suspect it can be parallelized to some extent; I can imagine schemes where one would pre-generate* the locations of points to be resolved, and then partition them between threads based on spatial proximity -- perhaps with sync points at specific locations of the point list. Quite unclear how much performance we could win though, as we're likely in the area of diminishing returns there.

The R* tree was more of a "what can we plug in in a short amount of time that more or less matches the requirements" rather than a particularly conscious choice. We tried kD-trees before, and they weren't that great due to high branching factors, resulting in pointer chasing and branching in execution. Perhaps there's something better out there :)

_\* Pre-generating the order of pixel locations for synthesis could enable us to use a high quality low-discrepancy sequence instead of just random, possibly improving the quality of generated images._

**Comment by Mr4k** [comment]
Oh interesting point about the pre generated sequences. I had thought about doing something like that but I was worried that it would change the distribution. However it sounds like a different distribution could potentially allow for higher quality (or at least more predictable) synthesis if chosen carefully.
